sdcm/utils/version_utils.py

```diff
@@ -4,65 +4,166 @@
+
+# gemini version 1.0.1, commit ef7c6f422c78ef6b84a6f3bccf52ea9ec846bba0, date 2019-05-16T09:56:16Z
+GEMINI_VERSION_RE = re.compile(r'\s(?P<gemini_version>([\d]+\.[\d]+\.[\d]+)?),')
+
 SCYLLA_VERSION_RE = re.compile(r"\d+(\.\d+)?\.[\d\w]+([.~][\d\w]+)?")
+PRIMARY_XML_GZ_REGEX = re.compile(r'="(.*?primary.xml.gz)"')
+
+REPOMD_XML_PATH = "repodata/repomd.xml"
+URL_REGEX_AND_FORMAT_LIST = [
+    (re.compile(r"baseurl=(?P<url>http.*scylladb.com.*)"), "{url}" + REPOMD_XML_PATH),
+    (re.compile(r"deb\s+(?P<url>http.*?)\s(?P<version_code_name>.*?)\s(?P<component>.*?)$"),
+     "{url}/dists/{version_code_name}/{component}/binary-amd64/Packages.gz"),
+    (re.compile(r"deb\s+\[arch=(?P<arch>.*?)\]\s(?P<url>http.*?)\s(?P<version_code_name>.*?)\s(?P<component>.*?)$"),
+     "{url}/dists/{version_code_name}/{component}/binary-{arch}/Packages"),
+]
+
+SCYLLA_URL_REGEX = re.compile(".*(?P<url>http.*scylladb.com.*).*")
+
+SCYLLA_URL_RESPONSE_TIMEOUT = 30
+SUPPORTED_XML_EXTENSIONS = ("xml", "xml.gz")
+SUPPORTED_FILE_EXTENSIONS = ("list", "repo", "Packages", "gz") + SUPPORTED_XML_EXTENSIONS
+BASE_ARCH_KEY = "$basearch"
+RELEASE_VERSION_KEY = "$releasever"
+SCYLLA_MANAGER_KEY = "scylla-manager"
+URL_ENV_VARIABLES = {
+    BASE_ARCH_KEY: ["x86_64", "noarch"],
+    RELEASE_VERSION_KEY: ["7"],
+    SCYLLA_MANAGER_KEY: ["x86_64"],
+}
+VERSION_NOT_FOUND_ERROR = "URL isn't a correct deb list or yum repo"
+
+
+def get_all_scylla_url_patterns(url):
+    response = requests.get(url=url)
+    response_data = response.text
+    if not response_data:
+        raise ValueError(f"The file that downloaded from url '{url}' is empty")
+    return list({line for line in response_data.split('\n') if SCYLLA_URL_REGEX.match(line)})
+
+
+def extract_all_urls_from_scylla_patterns(scylla_patterns, url_regex_details):
+    """
+    There are 4 different patterns:
+        1. "Empty" URL - The URL not contains any special variable and need to add the following suffix
+            "{url}repodata/repomd.xml"
+        2. "Basearch" URL - The URL contains the special variable "$basearch".
+            The value of the variable must be replaced by one of the supported values (URL_ENV_VARIABLES[BASE_ARCH_KEY])
+        3. "releasever" URL - The URL contains the special variable "$releasever".
+            The value of the variable must be replaced by one of the supported values (
+            URL_ENV_VARIABLES[RELEASE_VERSION_KEY])
+        4. Multiple special variables - The URL contains multiple special variables.
+            Each special variable replaced with supported values.
+    """
 
+    all_scylla_version_urls = set()
+    for url_regex, url_format in url_regex_details:
+        for pattern in scylla_patterns:
+            packages_details = url_regex.match(pattern)
+            if not packages_details:
+                continue
+
+            version_url = url_format.format(**packages_details.groupdict())
+            is_base_arch_found = BASE_ARCH_KEY in version_url
```

Comment:
the logic is too complex to grasp what's going on,

try something like that, using the built in template to help:
```python
from string import Template

all_scylla_version_urls = set()
...
# for scylla-manager we never a noarch
arch_list =  ["x86_64", "noarch"] if 'scylla-manager' not in version_url else ["x86_64"]

for arch in ["x86_64", "noarch"]:
     all_scylla_version_urls.add(Template(version_url).substitute(basearch=arch, releasever='7'))
```

also the excessive use of variable for each thing, make this very hard to follow, 
since you need to scroll back and forth to understand what is what.